### PR TITLE
add trycatch support

### PR DESCRIPTION
**Pull request by lightszero**
working

### CODE REVIEW

File: src/Neo.Compiler.MSIL/MSIL/Conv_Multi.cs [hunk 1]
```diff
@@ -1348,11 +1348,24 @@ private int _ConvertNewObj(OpCode src, NeoMethod to)
             }
             else if (_type.DeclaringType.FullName.Contains("Exception"))
             {
+                //对异常对象要留一个，因为catch 会处理这个栈
                 _Convert1by1(VM.OpCode.NOP, src, to);//空白
```
Comment (ShawnYun) [inline src/Neo.Compiler.MSIL/MSIL/Conv_Multi.cs]: If we use the optimizer to delete `NOP`, does that affect this?

Comment (shargon) [reply src/Neo.Compiler.MSIL/MSIL/Conv_Multi.cs]: #if DEBUG ?

Comment (lightszero) [reply src/Neo.Compiler.MSIL/MSIL/Conv_Multi.cs]: do not worry，optimzer is safe to delete any NOP

**Comment by Tommo-L** [comment]
@lightszero dalao, vm has merged the try-catch pr, we can continue this pr now. 😂

**Comment by lightszero** [comment]
go go go

**Comment by lightszero** [comment]
@Tommo-L  many unittest had fail, help

**Comment by Tommo-L** [comment]
> @Tommo-L many unittest had fail, help

Ok, I'll help fix these ut.

**Comment by Tommo-L** [comment]
This pr block by #219, as neo-core have merged the offset pr, most ut need to adjust.

File: src/Neo.Compiler.MSIL/MSIL/Converter.cs [hunk 2]
```diff
@@ -346,23 +346,77 @@ private void ConvertAddrInMethod(NeoMethod to)
...
+                        if (info.catch_Infos.Count != 1)
+                            throw new Exception("only support one catch for now.");
+                        var first = info.catch_Infos.First().Value;
+                        var bytesCatch = BitConverter.GetBytes(first.addrBegin);
+                        var buf = new byte[8];
+                        Array.Copy(bytesCatch, 0, buf, 0, 4);
+                        Array.Copy(bytesFinally, 0, buf, 4, 4);
+                        var trycode = Convert1by1(VM.OpCode.TRY_L, src, to, buf);
+                        trycode.needfix = true;
+
```
Comment (Tommo-L) [inline src/Neo.Compiler.MSIL/MSIL/Converter.cs:415]: So we don't support to nest try-catch? And this ut will fail.

```cs

public static object tryNest()
{
	int v = 0;
	try
	{
		try
		{
			v = 2;
			throwcall();
		}
		catch
		{
			v = 3;
			throwcall();
		}
		finally
		{
			throwcall();
			v++;
		}
	}
	catch
	{
		v++;
	}
	return v;
}
```

Comment (erikzhang) [reply src/Neo.Compiler.MSIL/MSIL/Converter.cs:415]: > So we don't support to nest try-catch?

In neo-vm, it supports. In the compiler, I don't know.

File: src/Neo.Compiler.MSIL/MSIL/Converter.cs [hunk 3]
```diff
@@ -346,23 +346,77 @@ private void ConvertAddrInMethod(NeoMethod to)
...
                     {
-                        var _addr = addrconv[c.srcaddr];
-                        Int32 addroff = (Int32)(_addr - c.addr);
-                        c.bytes = BitConverter.GetBytes(addroff);
-                        c.needfix = false;
+                        var srcCatch = BitConverter.ToInt32(c.bytes, 0);
+                        var srcFinal = BitConverter.ToInt32(c.bytes, 4);
+                        if (srcCatch == -1)
+                        {
+                            var bytesCatch = new byte[0, 0, 0, 0];
```
Comment (Tommo-L) [inline src/Neo.Compiler.MSIL/MSIL/Converter.cs:355]:
```suggestion
                            var bytesCatch = new byte[]{ 0, 0, 0, 0};
```

File: src/Neo.Compiler.MSIL/MSIL/Converter.cs [hunk 4]
```diff
@@ -346,23 +346,77 @@ private void ConvertAddrInMethod(NeoMethod to)
...
+                        else
+                        {
+                            var _addrCatch = addrconv[srcCatch];
+                            Int32 addroffCatch = (Int32)(_addrCatch - c.addr);
+                            var bytesCatch = BitConverter.GetBytes(addroffCatch);
+                            Array.Copy(bytesCatch, 0, c.bytes, 0, 4);
+                        }
+                        if (srcFinal == -1)
+                        {
+                            var bytesFinal = new byte[0, 0, 0, 0];
```
Comment (Tommo-L) [inline src/Neo.Compiler.MSIL/MSIL/Converter.cs:367]:
```suggestion
                            var bytesFinal = new byte[]{0, 0, 0, 0};
```

File: src/Neo.Compiler.MSIL/MSIL/Converter.cs [hunk 5]
```diff
@@ -346,23 +346,77 @@ private void ConvertAddrInMethod(NeoMethod to)
...
         {
+            //add try code
+            if (method.tryPositions.Contains(src.addr))
+            {
+                foreach (var info in method.tryInfos)
+                {
+                    if (info.addr_Try_Begin == src.addr)
+                    {
+                        var bytesFinally = BitConverter.GetBytes(info.addr_Finally_Begin);
+                        if (info.catch_Infos.Count != 1)
```
Comment (Tommo-L) [inline src/Neo.Compiler.MSIL/MSIL/Converter.cs:406]:
```suggestion
                        if (info.catch_Infos.Count > 1)
```

**Comment by Tommo-L** [comment]
Fixed by https://github.com/neo-project/neo-devpack-dotnet/pull/259

File: src/Neo.Compiler.MSIL/MSIL/Converter.cs [hunk 6]
```diff
@@ -346,23 +346,77 @@ private void ConvertAddrInMethod(NeoMethod to)
...
+                            throw new Exception("only support one catch for now.");
+                        var first = info.catch_Infos.First().Value;
+                        var bytesCatch = BitConverter.GetBytes(first.addrBegin);
+                        var buf = new byte[8];
+                        Array.Copy(bytesCatch, 0, buf, 0, 4);
+                        Array.Copy(bytesFinally, 0, buf, 4, 4);
+                        var trycode = Convert1by1(VM.OpCode.TRY_L, src, to, buf);
+                        trycode.needfix = true;
```
Comment (Tommo-L) [inline src/Neo.Compiler.MSIL/MSIL/Converter.cs:414]:
```suggestion
                        trycode.needfix = true;
                        break;
```

File: src/Neo.Compiler.MSIL/MSIL/Converter.cs (hunk 5 above)
Comment (Tommo-L) [reply src/Neo.Compiler.MSIL/MSIL/Converter.cs]: @shargon  we need to support `try-finally` without catch body.

File: src/Neo.Compiler.MSIL/MSIL/Conv_Multi.cs [hunk 7]
```diff
@@ -1308,21 +1308,34 @@ private int ConvertNewObj(ILMethod from, OpCode src, NeoMethod to)
...
+                // NeoVM `catch` instruction need one exception parameter
                 Convert1by1(VM.OpCode.NOP, src, to);
                 var pcount = _type.Parameters.Count;
-                for (var i = 0; i < pcount; i++)
+                if (pcount == 0) // If there is no parameter, insert one pararmeter
                 {
-                    Insert1(VM.OpCode.DROP, "", to);
+                    ConvertPushString("usererror", src, to);
+                }
+                else if (pcount == 1)
```
Comment (shargon) [inline src/Neo.Compiler.MSIL/MSIL/Conv_Multi.cs:1322]: != and remove the empty block

Comment (lightszero) [reply src/Neo.Compiler.MSIL/MSIL/Conv_Multi.cs:1322]: maybe add a comment for it.

**Comment by ProDog** [comment]
Can we supports catching details of exceptions? like:
```C#
try
{
                   
}

catch (Exception ex)
{
    Runtime.Notify("error:" + ex.Message);
}
```

**Comment by lightszero** [comment]
> Can we supports catching details of exceptions? like:
> 
> ```cs
> try
> {
>                    
> }
> 
> catch (Exception ex)
> {
>     Runtime.Notify("error:" + ex.Message);
> }
> ```
we can do that next step.

File: src/Neo.Compiler.MSIL/MSIL/Converter.cs [hunk 8]
```diff
@@ -346,23 +347,83 @@ private void ConvertAddrInMethod(NeoMethod to)
...
+                        }
+                        if (srcFinal == -1)
+                        {
+                            var bytesFinal = new byte[] { 0, 0, 0, 0 };
+                            Array.Copy(bytesFinal, 0, c.bytes, 4, 4);
+                        }
+                        else
+                        {
+                            var _addrFinal = addrconv[srcFinal];
+                            Int32 addroffFinal = (Int32)(_addrFinal - c.addr);
```
Comment (shargon) [inline src/Neo.Compiler.MSIL/MSIL/Converter.cs:374]:
```suggestion
                            int addroffFinal = (int)(_addrFinal - c.addr);
```

File: src/Neo.Compiler.MSIL/MSIL/Converter.cs [hunk 9]
```diff
@@ -346,23 +347,83 @@ private void ConvertAddrInMethod(NeoMethod to)
...
+                        }
                     }
-                    catch
+                    else
                     {
-                        throw new Exception("cannot convert addr in: " + to.name + "\r\n");
+                        try
+                        {
+                            var _addr = addrconv[c.srcaddr];
+                            Int32 addroff = (Int32)(_addr - c.addr);
```
Comment (shargon) [inline src/Neo.Compiler.MSIL/MSIL/Converter.cs:384]:
```suggestion
                            int addroff = (int)(_addr - c.addr);
```

**Comment by Tommo-L** [comment]
@lightszero need your help, I have tried to fix it but failed, I need more time. Could you have a look?

```cs
public static object tryWithTwo()
{
	int v = 0;
	try
	{
		try
		{
			v++;
		}
		catch
		{
			v += 2;
		}
		finally
		{
			v += 3;
		}
	}
	catch
	{
		v += 4;
	}
	finally
	{
		v += 5;
	}
	return v;
}
```
_Found by_ @ProDog

**Comment by lightszero** [comment]
got it,I will check that.
but the ut had passed?so ut had problem?

**Comment by Tommo-L** [comment]
It's a new ut. The current ut all are passed. Should I send a new PR to add the new ut.  And the source code is here, https://github.com/neo-project/neo-devpack-dotnet/pull/186#issuecomment-621716152

**Comment by lightszero** [comment]
you could direct add ut in this pr.

**Comment by Tommo-L** [comment]
Plz, merge this pr https://github.com/neo-project/neo-devpack-dotnet/pull/261, I don't have the authorization to push commit directly.

**Comment by lightszero** [comment]
@Tommo-L done,it worked.

**Comment by Tommo-L** [comment]
Need your last test @ProDog

**Comment by ProDog** [comment]
> Need your last test @ProDog

Cool, I have tested it passed.

**Pull request merged**